onnxruntime/core/providers/cuda/nn/conv.cc

```diff
@@ -349,66 +350,125 @@ Status Conv<T, NHWC>::UpdateState(OpKernelContext* context, bool bias_expected)
       CUDA_CALL_THROW(cudaMemsetAsync(s_.b_zero, 0, malloc_size, Stream(context)));
     }
 
-    if (!s_.cached_benchmark_results.contains(x_dims_cudnn)) {
-      // set math type to tensor core before algorithm search
-      if constexpr (std::is_same<T, MLFloat16>::value)
-        CUDNN_RETURN_IF_ERROR(cudnnSetConvolutionMathType(s_.conv_desc, CUDNN_TENSOR_OP_MATH));
-
-      cudnnConvolutionFwdAlgoPerf_t perf;
-      int algo_count = 1;
-      int cudnn_conv_algo = cuda_ep->GetCudnnConvAlgo();
-      ORT_ENFORCE(cudnn_conv_algo > -1 && cudnn_conv_algo < 3, "cudnn_conv_algo should be 0, 1 or 2, but got ", cudnn_conv_algo);
-      switch (cudnn_conv_algo) {
-        case 0: {
-          static constexpr int num_algos = CUDNN_CONVOLUTION_FWD_ALGO_COUNT;
-          size_t max_ws_size = cuda_ep->GetCudnnConvUseMaxWorkspace() ? GetMaxWorkspaceSize(GetCudnnHandle(context), s_, kAllAlgos, num_algos)
-                                                                      : AlgoSearchWorkspaceSize;
-          // Use GetTransientScratchBuffer() so the workspace can be freed instead of cached.
-          // Because the benchmarking uses a huge amount of memory, e.g. a few GBs.
-          IAllocatorUniquePtr<void> algo_search_workspace = GetTransientScratchBuffer<void>(max_ws_size);
-          CUDNN_RETURN_IF_ERROR(cudnnFindConvolutionForwardAlgorithmEx(
-              GetCudnnHandle(context),
-              s_.x_tensor,
-              s_.x_data,
-              s_.w_desc,
-              s_.w_data,
-              s_.conv_desc,
-              s_.y_tensor,
-              s_.y_data,
-              1,            // requestedAlgoCount
-              &algo_count,  // returnedAlgoCount
-              &perf,
-              algo_search_workspace.get(),
-              max_ws_size));
-          break;
+    if (NHWC && s_.cudnn_fe_graph == nullptr) {
+      s_.cudnn_fe_graph = std::make_unique<cudnn_frontend::graph::Graph>();
+
+      cudnn_frontend::DataType_t data_type = CudnnFeTensor::GetDataType<CudaT>();
+
+      s_.cudnn_fe_graph->set_io_data_type(data_type)
+          .set_compute_data_type(data_type == cudnn_frontend::DataType_t::HALF ? cudnn_frontend::DataType_t::FLOAT : data_type)
+          .set_intermediate_data_type(data_type);
+
+      s_.cudnn_fe_X = s_.cudnn_fe_graph->tensor(CudnnFeTensor(X, "x", NHWC, data_type).Get());
+      s_.cudnn_fe_W = s_.cudnn_fe_graph->tensor(CudnnFeTensor(W, "w", NHWC, data_type).Get());
+
+      std::vector<int64_t> fe_conv_padding;
+      std::vector<int64_t> fe_conv_stride;
+      std::vector<int64_t> fe_conv_dilations;
+      for (size_t i = 0; i < kernel_shape.size(); i++) {
```

Comment:
Using a for loop might be more efficient since loop control is done only once... but from a readability standpoint std::copy might be better.

What is the type of pads/strides/diliations so that a direct copy cannot be used? In theory this also leads to a memory corruption since you assume that all 3 input arrays have the correct size which is potentially not true.

Reply:
nitpick, from a memory allocation standpoint I am curious of a fixed size array of 4? would be sufficient... if cudnn_
frontend required std::vector instead of e.g. gls::span or it_begin, it_end (which is a gls::span).

Reply:
Cudnn frontend requires a std::vector and checks the validity of the convolution based on the vector lengths. 

@gedoensmax setting a constant size of 6 for the vectors gives this error:

````
!!!! nb_spatial_dims > nb_tensor_dims - 2; /dvs/p4/build/sw/gpgpu/MachineLearning/cudnn/src/graph/src/convolution_ops.cpp:999
err "CUDNN_STATUS_BAD_PARAM" (2000) received on "is_valid_convolution(xDesc, wDesc, cDesc, yDesc)"
err "CUDNN_STATUS_BAD_PARAM" (2000) received on "convolution.init(xDesc, wDesc, cDesc, yDesc)"
err "CUDNN_STATUS_BAD_PARAM" (2000) received on "finalize_internal()"

E! CuDNN (v90000) function cudnnBackendFinalize() called:
e!     Info: Traceback contains 4 message(s)
e!         Error: CUDNN_STATUS_BAD_PARAM; Reason: nb_spatial_dims > nb_tensor_dims - 2
e!         Error: CUDNN_STATUS_BAD_PARAM; Reason: is_valid_convolution(xDesc, wDesc, cDesc, yDesc)
e!         Error: CUDNN_STATUS_BAD_PARAM; Reason: convolution.init(xDesc, wDesc, cDesc, yDesc)
e!         Error: CUDNN_STATUS_BAD_PARAM; Reason: finalize_internal()
e! Time: 2024-02-02T10:15:09.274254 (0d+0h+0m+0s since start)
e! Process=280080; Thread=280080; GPU=NULL; Handle=NULL; StreamId=NULL.

````

Solution from my side would be this (is const_t kernel_shape_size more efficient than using 3 times .size()?):
````

size_t kernel_shape_size = kernel_shape.size();
auto conv_options = cudnn_frontend::graph::Conv_fprop_attributes()
              .set_padding(std::vector<int64_t>(pads.begin(), pads.begin() + kernel_shape_size))
              .set_stride(std::vector<int64_t>(strides.begin(), strides.begin() + kernel_shape_size))
              .set_dilation(std::vector<int64_t>(dilations.begin(), dilations.begin() + kernel_shape_size));
````